extra_data/keydata.py

```diff
@@ -133,31 +142,26 @@ def split_trains(self, parts=None, trains_per_part=None):
         for s in split_trains(len(self.train_ids), parts, trains_per_part):
             yield self.select_trains(s)
 
-    def data_counts(self):
+    def data_counts(self, labelled=True):
         """Get a count of data entries in each train.
 
-        Returns a pandas series with an index of train IDs.
+        If *labelled* is True, returns a pandas series with an index of train
+        IDs. Otherwise, returns a NumPy array of counts to match ``.train_ids``.
         """
-        import pandas as pd
-
-        seq_series = []
-
-        for f in self.files:
-            if self.section == 'CONTROL':
-                counts = np.ones(len(f.train_ids), dtype=np.uint64)
-            else:
-                _, counts = f.get_index(self.source, self._key_group)
+        train_ids = np.concatenate([c.train_ids for c in self._data_chunks])
+        counts = np.concatenate([c.counts for c in self._data_chunks])
 
-            if self.inc_suspect_trains:
-                s = pd.Series(counts, index=f.train_ids)
-            else:
-                s = pd.Series(counts[f.validity_flag], index=f.valid_train_ids)
-            seq_series.append(s)
-
-        ser = pd.concat(sorted(seq_series, key=lambda s: s.index[0]))
-        # Select out only the train IDs of interest
-        train_ids = ser.index.intersection(np.asarray(self.train_ids))
-        return ser.loc[train_ids]
+        if labelled:
+            import pandas as pd
+            return pd.Series(counts, index=train_ids)
+        else:
+            # self.train_ids is always sorted. The train IDs from chunks
+            # should be in order, but sometimes trains are written out of order.
+            # Reorder the counts to match self.train_ids.
+            assert len(train_ids) == len(self.train_ids)
```

Comment:
Wouldn't this fail after dropping missing trains?

Reply:
`self._data_chunks` should have chunks contaiing only the selected trains. It wouldn't hurt to add a test for it, though.

Reply:
I added a test :slightly_smiling_face:

Reply:
Thank you